Virtual node crashing with error  invalid memory address

#### What happened:

I have peered AKS cluster with on-premise K8S cluster and was working well, suddenly virtual pod started crashing with below error.
Can you help me if I'm doing something wrong.


```
E1024 19:50:03.169332       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 130 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1c67920?, 0x3358dc0})
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/runtime/runtime.go:74 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002687780?})
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/runtime/runtime.go:48 +0x75
panic({0x1c67920, 0x3358dc0})
        /usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/api/core/v1.(*Pod).GetObjectKind(0x1ef74a0?)
        <autogenerated>:1
k8s.io/client-go/tools/reference.GetReference(0x43?, {0x23f8ac8, 0x0})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/reference/ref.go:59 +0x137
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc0005c0700, {0x23f8ac8?, 0x0}, 0x1?, {0x1f2255e, 0x7}, {0x1f30546, 0x10}, {0xc000e75d60, 0x4c})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/record/event.go:330 +0x79
k8s.io/client-go/tools/record.(*recorderImpl).Event(0x1fa6797?, {0x23f8ac8?, 0x0?}, {0x1f2255e?, 0x1?}, {0x1f30546?, 0x0?}, {0xc000e75d60?, 0xc000458000?})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/record/event.go:355 +0x4f
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/workload.(*NamespacedPodReflector).Handle(0xc000d09440, {0x240e8a8, 0xc001911110}, {0xc00268b590, 0x4})
        /tmp/builder/pkg/virtualKubelet/reflection/workload/podns.go:127 +0x5ca
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).handle(0xc000458cb0, {0x240e838, 0xc000056030}, {{0xc000656540?, 0x8?}, {0xc00268b590?, 0x0?}})
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:234 +0x9c7
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).processNextWorkItem(0xc000458cb0)
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:190 +0xf1
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).runWorker(0xc000c0aea0?)
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:167 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x23f22c0, 0xc000b42540}, 0x1, 0xc0000c4a80)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:90 +0x25
created by github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).Start
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:94 +0x1ae
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbfa200]

goroutine 130 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002687780?})
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x1c67920, 0x3358dc0})
        /usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/api/core/v1.(*Pod).GetObjectKind(0x1ef74a0?)
        <autogenerated>:1
k8s.io/client-go/tools/reference.GetReference(0x43?, {0x23f8ac8, 0x0})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/reference/ref.go:59 +0x137
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc0005c0700, {0x23f8ac8?, 0x0}, 0x1?, {0x1f2255e, 0x7}, {0x1f30546, 0x10}, {0xc000e75d60, 0x4c})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/record/event.go:330 +0x79
k8s.io/client-go/tools/record.(*recorderImpl).Event(0x1fa6797?, {0x23f8ac8?, 0x0?}, {0x1f2255e?, 0x1?}, {0x1f30546?, 0x0?}, {0xc000e75d60?, 0xc000458000?})
        /go/pkg/mod/k8s.io/client-go@v0.23.6/tools/record/event.go:355 +0x4f
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/workload.(*NamespacedPodReflector).Handle(0xc000d09440, {0x240e8a8, 0xc001911110}, {0xc00268b590, 0x4})
        /tmp/builder/pkg/virtualKubelet/reflection/workload/podns.go:127 +0x5ca
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).handle(0xc000458cb0, {0x240e838, 0xc000056030}, {{0xc000656540?, 0x8?}, {0xc00268b590?, 0x0?}})
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:234 +0x9c7
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).processNextWorkItem(0xc000458cb0)
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:190 +0xf1
github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).runWorker(0xc000c0aea0?)
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:167 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x23f22c0, 0xc000b42540}, 0x1, 0xc0000c4a80)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
        /go/pkg/mod/k8s.io/apimachinery@v0.23.6/pkg/util/wait/wait.go:90 +0x25
created by github.com/liqotech/liqo/pkg/virtualKubelet/reflection/generic.(*reflector).Start
        /tmp/builder/pkg/virtualKubelet/reflection/generic/reflector.go:94 +0x1ae
```
#### What you expected to happen:

#### How to reproduce it (as minimally and precisely as possible):
Create a Pod directly in remote cluster(in my case AKS), then local cluster(on-premise) virtual node starts with above error.


#### Anything else we need to know?:

#### Environment:
- Liqo version: 0.5.4
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Network plugin and version:
- Install tools:
- Others:

--- a/pkg/virtualKubelet/reflection/workload/podns.go
+++ b/pkg/virtualKubelet/reflection/workload/podns.go
@@ -124,7 +124,7 @@ func (npr *NamespacedPodReflector) Handle(ctx context.Context, name string) erro
 
 	// Abort the reflection if the remote object is not managed by us, as we do not want to mutate others' objects.
 	if (remoteExists && !forge.IsReflected(remote)) || (shadowExists && !forge.IsReflected(shadow)) {
-		if !localExists { // Do not output the warning event in case the event was triggered by the remote object (i.e., the local one does not exists).
+		if localExists { // Do not output the warning event in case the event was triggered by the remote object (i.e., the local one does not exists).
 			klog.Infof("Skipping reflection of local pod %q as remote already exists and is not managed by us", npr.LocalRef(name))
 			npr.Event(local, corev1.EventTypeWarning, forge.EventFailedReflection, forge.EventFailedReflectionAlreadyExistsMsg())
 		}
@@ -135,7 +135,7 @@ func (npr *NamespacedPodReflector) Handle(ctx context.Context, name string) erro
 	// The local pod does no longer exist. Ensure the shadowpod is absent from the remote cluster.
 	if !localExists {
 		defer tracer.Step("Ensured the absence of the remote object")
-		if !kerrors.IsNotFound(serr) {
+		if shadowExists {
 			klog.V(4).Infof("Deleting remote shadowpod %q, since local pod %q does no longer exist", npr.RemoteRef(name), npr.LocalRef(name))
 			return npr.DeleteRemote(ctx, npr.remoteShadowPodsClient, "ShadowPod", name, shadow.GetUID())
 		}

--- a/pkg/virtualKubelet/reflection/workload/podns_test.go
+++ b/pkg/virtualKubelet/reflection/workload/podns_test.go
@@ -94,6 +94,22 @@ var _ = Describe("Namespaced Pod Reflection Tests", func() {
 				err           error
 			)
 
+			WhenBodyRemoteNotManagedByReflection := func() func() {
+				return func() {
+					var shadowBefore *vkv1alpha1.ShadowPod
+
+					BeforeEach(func() {
+						shadowBefore = CreateShadowPod(liqoClient, &shadow)
+					})
+
+					It("should succeed", func() { Expect(err).ToNot(HaveOccurred()) })
+					It("the remote object should be unmodified", func() {
+						shadowAfter := GetShadowPod(liqoClient, RemoteNamespace, PodName)
+						Expect(shadowAfter).To(Equal(shadowBefore))
+					})
+				}
+			}
+
 			BeforeEach(func() {
 				local = corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: PodName, Namespace: LocalNamespace}}
 				remote = corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: PodName, Namespace: RemoteNamespace}}
@@ -124,6 +140,7 @@ var _ = Describe("Namespaced Pod Reflection Tests", func() {
 
 				When("the remote object does not exist", WhenBody(false))
 				When("the remote object does exist", WhenBody(true))
+				When("the remote object does exist, but is not managed by the reflection", WhenBodyRemoteNotManagedByReflection())
 			})
 
 			When("the local object does exist and is not terminating", func() {
@@ -232,19 +249,7 @@ var _ = Describe("Namespaced Pod Reflection Tests", func() {
 					})
 				})
 
-				When("the remote object already exists, but is not managed by the reflection", func() {
-					var shadowBefore *vkv1alpha1.ShadowPod
-
-					BeforeEach(func() {
-						shadowBefore = CreateShadowPod(liqoClient, &shadow)
-					})
-
-					It("should succeed", func() { Expect(err).ToNot(HaveOccurred()) })
-					It("the remote object should be unmodified", func() {
-						shadowAfter := GetShadowPod(liqoClient, RemoteNamespace, PodName)
-						Expect(shadowAfter).To(Equal(shadowBefore))
-					})
-				})
+				When("the remote object already exists, but is not managed by the reflection", WhenBodyRemoteNotManagedByReflection())
 
 				When("the local object has already the appropriate offloading label", func() {
 					BeforeEach(func() {

